### ISSUE TITLE
[Vision Tracker] WebdriverIO Fiddle Platform

### ISSUE THREAD

**Issue by christian-bromann** [webdriverio/webdriverio#4709]
The team has the following requirements to such an app:

- create fiddle projects that are shareable (maybe reusing StackBlitz)
- allow to run commands synchronously

```js
browser.url('https://webdriver.io');
const title = browser.getTitle();
assert.equal(title, 'WebdriverIO · Next-gen WebDriver test framework for Node.js');
```

- check if we can bring fibers to the browser
- otherwise run it from backend
- have a file storage for every project on S3 where fiddle code is being stored
- one window (left) with wdio test code, split view right with stdout (top) and WebdriverIO logs on the bottom
- if commands can't be run in sync built minimal backend that runs the code and returns the logs and stdout
- UI components (Material UI) for Share button, Run button, select environments (OS, browser, version)
- download button to download test suite and run as wdio test suite

**Comment by mgrybyk** [comment]
Is it related to https://github.com/webdriverio/webdriverio/issues/4503 ?

**Comment by christian-bromann** [comment]
> Is it related to #4503 ?

Yes, I was talking with @baruchvlz today and that would be a great new initiative we could work on (given that we are in the same office and have a hackathon coming up). I looked into Kevins and Wills projects and both are great. In this issue I wanted to formulize what we think could be a good first POC.

**Comment by baruchvlz** [comment]
### WDIO Fiddle PoC

**Stack**
- new mono-repo inside webdriverio org
- NextJS - Server-side rendering for ReactJS
    - By having this separate from the electron wrapper we can also use it to spin it off locally on a browser window via the CLI:

        $ wdio fiddle [flags]

- Electron - desktop application wrapper for JavaScript SPAs
- MobX as a state manager
    - This is a lot easier to use than Redux, especially since we'll have to do state updates in the WDIO hooks
- WebdriverIO as the test execution framework.

**What?**

Provide a GUI where users can easily write their own tests, or listen to file changes in a directory, and have real-time updates on these tests as they are executed.

What to display

- Browser window
    - Injected iframe url where user can see in real time the test executions
- Logger viewers
    - two tabs [Test logs / Console Logs]
    - Test Logs tab
        - action name - test name - execution time
        - in progress / failed / passed
            - this will be done via colours in the list item
        - sub actions like click, wait for, set value
    - Console log tab
        - any console logs from the user by overwriting the window.console global
        - nice to have - colours for different methods
- [Optional] Text editor

    **Marked optional because we'll want the user to also have the ability to listen to local files and folders and refresh tests as changes happen**

    - For PoC we won't have multiple files / folders support, just one file where the user can quickly write tests

**How?**

We will utilise WebdriverIO as the test execution framework and browser communicator in order to provide an easier test creation and report visualisation for the user. Utilising MobX as a state manager we will be able to display this information by doing state updates in the [WDIO hooks](https://webdriver.io/docs/options.html#hooks). 

Hooks to use

- onPrepare
- before/after Suite
- before/after Test
- before/after Command
- onComplete

**Why?**

There are libraries that already support this and have gained popularity due to ease of use. By adding this to WDIO we can provide all the functionality of the library in a simple to easy, straight forward method that will attract more users.

**Questions**

- How can WDIO run the tests against the GUI's embedded (via iframe) browser view?
    - Verify if running locally or via electron
        - Locally → point config to local browser
        - Electron → point config to electron window
- Will it be possible to update react-mobx state inside WDIO hooks?

**Comment by christian-bromann** [comment]
Moving into a GitHub discussion: #6277
